File: go/sbgnvalidator.go
// Package sbgnvalidator validates SBGN-ML with built-in or explicit Schematron rules.
package sbgnvalidator

import (
	"context"

	"github.com/cannin/sbgn_validator/go/internal/rules"
	"github.com/cannin/sbgn_validator/go/internal/validator"
)

// Report is the normalized validation result.
type Report = validator.Report

// RuleInfo is provenance for the built-in ruleset.
type RuleInfo = rules.Info

// RulesInfo is provenance for the rules embedded in the Go package and binary.
func RulesInfo() (RuleInfo, error) {
	return rules.RulesInfo()
}

// Validate uses the built-in schema selected from the document language.
func Validate(ctx context.Context, documentPath, phase string) (Report, error) {
	schema, err := validator.CompileBuiltin(documentPath, phase)
	if err != nil {
		return Report{}, err
	}
	return schema.Validate(ctx, documentPath)
}

// ValidateWithRules uses an explicit Schematron file instead of built-in rules.
func ValidateWithRules(
	ctx context.Context,
	documentPath string,
	rulesPath string,
	phase string,
) (Report, error) {
	schema, err := validator.CompileSchema(rulesPath, phase)
	if err != nil {
		return Report{}, err
	}
	return schema.Validate(ctx, documentPath)
}

File: java/src/main/java/org/sbgn/schematron/SbgnValidator.java
package org.sbgn.schematron;

import java.nio.file.Path;

/** User-facing validation API with packaged rules as the default. */
public final class SbgnValidator {
    private SbgnValidator() {
    }

    /** Validate with the built-in schema selected from the document language. */
    public static CanonicalReport validate(Path document) throws Exception {
        return validate(document, null, "basic", NamespacePolicy.STRICT_03);
    }

    /** Validate with built-in rules or an explicit Schematron override. */
    public static CanonicalReport validate(
            Path document, Path rulesPath, String phase, NamespacePolicy namespacePolicy)
            throws Exception {
        BuiltinRules.DocumentInfo documentInfo = BuiltinRules.inspectDocument(document);
        String effectiveNamespace = namespacePolicy.effectiveNamespace(documentInfo.namespace());
        ReferenceValidator validator = rulesPath == null
                ? ReferenceValidator.builtinForDocument(
                        documentInfo.language(), phase, namespacePolicy, effectiveNamespace)
                : ReferenceValidator.forDocument(
                        rulesPath, phase, namespacePolicy, effectiveNamespace);
        return validator.validate(document);
    }

    /** Return provenance for the built-in rules. */
    public static RulesInfo rulesInfo() throws Exception {
        return BuiltinRules.rulesInfo();
    }
}
